From 6ccdf6970cf6d55b31d238f8ff878c605af464c8 Mon Sep 17 00:00:00 2001
From: Charles Ci-Jyun Wu <dminus@andestech.com>
Date: Mon, 17 Apr 2023 15:19:13 +0800
Subject: [PATCH 039/168] soc: andes: injection: Andes support injection
 feature

The injection feature utilizes reusable reserved memory
specified in the DTS. This allows shell scripts to be loaded into
the injection region either by loading files directly in QEMU or
restoring them to a specific location using gdb.
Linux then executes these scripts using rc.sysinit.

Reformed from the following patches on RISCV-Linux-5.4:
- (87831dd706c4) Fix: inject feature
- (62754f10ae3e) Fix injection feature
- (42d499327dcc) Fix: inject print message
- (6bb3adf3d930) hijack rename to inject

Reformed from the following patches on RISCV-Linux-6.1:
- soc: andes: injection: upgrade struct file_operations to struct proc_ops
- soc: andes: injection: injection script should adhere to the rules of
  the shell script

Signed-off-by: Charles Ci-Jyun Wu <dminus@andestech.com>
---
 drivers/soc/andes/Makefile      |  1 +
 drivers/soc/andes/inject_init.c | 64 +++++++++++++++++++++++++++++++++
 2 files changed, 65 insertions(+)
 create mode 100644 drivers/soc/andes/inject_init.c

diff --git a/drivers/soc/andes/Makefile b/drivers/soc/andes/Makefile
index fc25ab9c031f..1d7eb07ddb57 100644
--- a/drivers/soc/andes/Makefile
+++ b/drivers/soc/andes/Makefile
@@ -2,4 +2,5 @@
 # Copyright (C) 2023 Andes Technology Corporation.
 
 obj-y	+= andes_sbi.o
+obj-y	+= inject_init.o
 obj-$(CONFIG_ANDES_PPMA)	+= ppma.o
diff --git a/drivers/soc/andes/inject_init.c b/drivers/soc/andes/inject_init.c
new file mode 100644
index 000000000000..b46236afa6eb
--- /dev/null
+++ b/drivers/soc/andes/inject_init.c
@@ -0,0 +1,64 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * Copyright (C) 2020 Andes Technology Corporation.
+ */
+
+#include <linux/io.h>
+#include <linux/fs.h>
+#include <linux/init.h>
+#include <linux/proc_fs.h>
+#include <linux/seq_file.h>
+
+#ifdef CONFIG_PROC_FS
+
+// FIXME: make this FPGA-depedent
+#define INJECT_START    0x100000
+#define INJECT_SIZE     (0x200000 - INJECT_START)
+
+static void *inject;
+
+static int inject_show(struct seq_file *m, void *v)
+{
+	int i, pos;
+	unsigned char *p = (unsigned char *)inject;
+
+	if (inject != NULL && p[0] == 0) {
+		/* do nothing */
+	} else if (inject != NULL) {
+		for (i = 0, pos = 0; i < (INJECT_SIZE / PAGE_SIZE);
+		     i++, pos += PAGE_SIZE)
+			seq_write(m, (void *)&p[pos], PAGE_SIZE);
+	} else {
+		seq_printf(m, "#!/bin/sh\necho not found: ioremap failed.\n");
+	}
+	return 0;
+}
+
+static int inject_open(struct inode *inode, struct file *file)
+{
+	return single_open(file, inject_show, NULL);
+}
+
+static const struct proc_ops inject_fops = {
+	.proc_open	= inject_open,
+	.proc_read	= seq_read,
+	.proc_lseek	= seq_lseek,
+	.proc_release	= single_release,
+};
+
+static int __init inject_init(void)
+{
+	unsigned char *p;
+
+	inject = (void *)ioremap((phys_addr_t)INJECT_START, INJECT_SIZE);
+	p = (unsigned char *)inject;
+
+	if (p[0] == '#' && p[1] == '!')
+		proc_create("inject", 0444, NULL, &inject_fops);
+	else
+		iounmap(inject);
+
+	return 0;
+}
+arch_initcall(inject_init);
+#endif /* CONFIG_PROC_FS */
-- 
2.34.1

